#' Predict driver transcrition factors for scRNA-seq clusters
#'
#' Predict driver transcrition factors for scRNA-seq clusters based on TF ChIP-seq peaks from CistromeDB. 
#' To run this function, you must first install the rabit package (see from http://rabit.dfci.harvard.edu/). 
#'
#' @docType methods
#' @name RNAAnnotateTranscriptionFactor
#' @rdname RNAAnnotateTranscriptionFactor
#'
#' @param RNA Seurat object generated by \code{\link{RNARunSeurat}} function.
#' @param genes Data frame of differential expressed genes, generated by \code{\link{RNARunSeurat}}, \code{\link{FindMarkersMAESTRO}} or \code{\link{FindAllMarkersMAESTRO}} function.
#' @param cluster The cluster of given differential genes. 
#' For example, if the diff-genes are genrated using cluster 1,2,3 vs cluster 4,5,6, 
#' the cluster should be set as c(1,2,3,4,5,6).
#' @param project Output project name for the transcription factor analysis result. Default is the project name of the Seurat object.
#' @param method Use RABIT or LISA to identify the driver regulators. If Rabit is selected, should provide rabit annotation at the
#' same time. LISA do not need annotations for this step. Default is Rabit.
#' @param rabit.path Path of the rabit annotation files. The rabit annoation for CistromeDB is available at https://github.com/liulab-dfci/MAESTRO.
#' Download and unzip the annotation, supply the path of the annotations to \code{rabit.path}.
#' @param lisa.mode Mode to Run LISA, "local" or "web". If the mode is set as "local", 
#' please install LISA (https://github.com/qinqian/lisa) and download pre-computed datasets
#' following the instructions. The "web" mode is to run online version of LISA. 
#' In consideration of the connection issue, if users have mutiple clusters of differential genes,
#' the "local" mode is recommended.
#' @param conda.dir Directory where miniconda or anaconda is installed (required if the \code{lisa.mode} is set to "local"). 
#' For example, /home1/wangchenfei/miniconda3.
#' @param lisa.envname The name of the lisa conda environment (required if the \code{lisa.mode} is set to "local"). 
#' Default is "lisa".
#' @param organism Organism for the dataset. Only support "GRCh38" and "GRCm38". Default is "GRCh38".
#' @param top.tf Output the target genes for the top TFs. Default is 10.
#'
#' @author Dongqing Sun, Chenfei Wang
#'
#' @return A list of top enriched transcription factor family for each cluster. For each listed TF family,
#' TFs are ranked according to expression.
#'
#' @examples
#' data(pbmc.RNA)
#' data(human.immune.CIBERSORT)
#' pbmc.RNA.res <- RNARunSeurat(inputMat = pbmc.RNA, project = "PBMC.scRNA.Seurat", min.c = 10, min.g = 100, dims.use = 1:15)
#' pbmc.RNA.res$RNA <- RNAAnnotateCelltype(pbmc.RNA.res$RNA, pbmc.RNA.res$genes, human.immune.CIBERSORT, min.score = 0.05)
#' pbmc.tfs <- RNAAnnotateTranscriptionFactor(pbmc.RNA.res$RNA, pbmc.RNA.res$genes, project = "PBMC.scRNA.TF", rabit.path = "/homes/cwang/annotations/rabit")
#' pbmc.tfs
#'
#' @importFrom Seurat GetAssayData Idents
#' @export


RNAAnnotateTranscriptionFactor <- function(RNA, genes, cluster = NULL, project = RNA@project.name, 
                                           method = "RABIT", rabit.path, lisa.mode = "local", conda.dir = "", 
                                           lisa.envname = "lisa", organism = "GRCh38", top.tf = 10)
{
  if(organism == "GRCh38"){
    org = "hsa"
    data(human.tf.family)
    tf_family_list <- HUMAN.TFFamily}
  if(organism == "GRCm38"){
    org = "mmu"
    data(mouse.tf.family)
    tf_family_list <- MOUSE.TFFamily}

  if("cluster" %in% colnames(genes)){
    #genes$cluster <- as.factor(genes$cluster)
    genes <- genes[genes$avg_logFC>0,]
    ifAllcluster = TRUE
  }else{
    ifAllcluster = FALSE
    if(is.null(cluster)){
      cluster = "unknown"
      warning("If you need to filter TFs by expression, please specify the cluster.")
    }
    genes$cluster <- paste(cluster, collapse = ",")
    genes$gene <- rownames(genes)
  }

  method = toupper(method)
  if(method == "RABIT"){
    scorename = "log(Rabitscore)"
    out_fdr_max_log = tryCatch(expr = RunRABIT(genes = genes, project = project, rabit.path = rabit.path, organism = org),
                               error = function(cond) {return(cond$message)})
  }
  if(method == "LISA"){
    scorename = "log(Lisascore)"
    if(lisa.mode == "local"){
      out_fdr_max_log = RunLISALocal(genes = genes, project = project, organism = organism, conda.dir = conda.dir, lisa.envname = lisa.envname)
    }
    if(lisa.mode == "web"){
      out_fdr_max_log = RunLISAWeb(genes = genes, project = project, organism = organism)
    }
  }

  if(class(out_fdr_max_log) == "character"){
    message(out_fdr_max_log)
    tfList <- NULL
  }else{
    cluster_drivertf_list <- lapply(colnames(out_fdr_max_log), function(x){
      return(data.frame(row.names = rownames(out_fdr_max_log)[order(out_fdr_max_log[,x],decreasing=T)], factor = rownames(out_fdr_max_log)[order(out_fdr_max_log[,x],decreasing=T)], score = sort(out_fdr_max_log[,x],decreasing=T), stringsAsFactors = FALSE))
    })
    names(cluster_drivertf_list) <- colnames(out_fdr_max_log)


    if(ifAllcluster){
      cluster_cell_list = split(names(Idents(RNA)), Idents(RNA))
    }else{
      cluster_cell_list = list()
      if(length(cluster) == 1 && cluster == "unknown"){
        cluster_cell_list[["unknown"]] = names(Idents(RNA))
      }else{
        cluster_cell_list[[paste(cluster, collapse = ",")]] = names(Idents(RNA))[Idents(RNA) %in% cluster]
        cluster_cell_list[["others"]] = names(Idents(RNA))[!(Idents(RNA) %in% cluster)]
      }
    }

    cluster_avg_expr <- sapply(names(cluster_cell_list), function(x){
      return(Matrix::rowMeans(GetAssayData(object = RNA)[, cluster_cell_list[[x]]]))
    })

    cluster_tf_list_filter = sapply(names(cluster_drivertf_list), function(x){
      tf_family_filter = sapply(cluster_drivertf_list[[x]][,1], function(y){
        if(y %in% names(tf_family_list) & length(intersect(tf_family_list[[y]], rownames(cluster_avg_expr))) > 1){
          tf_family_expr = cluster_avg_expr[intersect(tf_family_list[[y]], rownames(cluster_avg_expr)),x]
          tf_family_expr = tf_family_expr[which(tf_family_expr != 0.00)]
          tf_family = names(tf_family_expr)[order(tf_family_expr, decreasing=T)]
          return(list(tf_family,cluster_drivertf_list[[x]][y,2]))
        }else{
          if(!(y %in% rownames(cluster_avg_expr)) || cluster_avg_expr[y,x] == 0.00){
            return(NULL)
          }else{
            return(list(y,cluster_drivertf_list[[x]][y,2]))
          }
        }
      }, simplify = FALSE)
      if(length(which(sapply(tf_family_filter,is.null))) > 0){
        tf_family_filter = tf_family_filter[-which(sapply(tf_family_filter,is.null))]
      }
      tf_family_filter_tf = sapply(tf_family_filter,function(xx){
        return(xx[[1]])
      })
      tf_family_filter_score = unlist(sapply(tf_family_filter,function(xx){
        return(xx[[2]])
      }))

      tf_family_filter_dedup_tf = tf_family_filter_tf[!duplicated(tf_family_filter_tf)]
      tf_family_filter_dedup_score = tf_family_filter_score[!duplicated(tf_family_filter_tf)]
      listlen = sapply(tf_family_filter_dedup_tf, function(xx){
        length(xx)
      })

      tf_family_filter_desubset_tf = sapply(tf_family_filter_dedup_tf,function(xx){
        ifsubset = sapply(tf_family_filter_dedup_tf, function(yy){
          all(xx %in% yy)
        })
        return(tf_family_filter_dedup_tf[ifsubset][[which.max(listlen[ifsubset])]])
      })
      tf_family_filter_desubset_score = sapply(tf_family_filter_dedup_tf,function(xx){
        ifsubset = sapply(tf_family_filter_dedup_tf, function(yy){
          all(xx %in% yy)
        })
        return(max(tf_family_filter_dedup_score[ifsubset]))
      })

      tf_family_filter_desubset_dedup_tf = tf_family_filter_desubset_tf[!duplicated(tf_family_filter_desubset_tf)]
      tf_family_filter_desubset_dedup_score = tf_family_filter_desubset_score[!duplicated(tf_family_filter_desubset_tf)]
      tf_family_filter_desubset_dedup_tf_str = lapply(tf_family_filter_desubset_dedup_tf, function(xx){
        return(paste(xx, collapse = " | "))
      })
      return(list(tf = unlist(tf_family_filter_desubset_dedup_tf_str)[1:top.tf], score = tf_family_filter_desubset_dedup_score[1:top.tf]))
    })

    cluster_tf_list_filter_tf = cluster_tf_list_filter["tf",]
    cluster_tf_list_filter_score = cluster_tf_list_filter["score",]
    if(length(cluster_tf_list_filter_tf) == 1){
      names(cluster_tf_list_filter_tf) = colnames(cluster_tf_list_filter)
      names(cluster_tf_list_filter_score) = colnames(cluster_tf_list_filter)
    }
    
    if(!ifAllcluster){
      names(cluster_tf_list_filter_tf) = paste(cluster, collapse = ",")
      names(cluster_tf_list_filter_score) = paste(cluster, collapse = ",")
    }

    cluster_tf_df = reshape2::melt(cluster_tf_list_filter_tf)[,c(2,1)]
    colnames(cluster_tf_df) = c("Cluster","TF")
    cluster_score_df = reshape2::melt(cluster_tf_list_filter_score)[,c(2,1)]
    colnames(cluster_score_df) = c("Cluster",scorename)
    cluster_tf_df[,scorename] = round(cluster_score_df[, scorename], 2)

    if("assign.ident" %in% colnames(RNA@meta.data)){
      if(ifAllcluster){
        reg_table = data.frame(Cluster = Idents(RNA), CelltypeAnnotation = RNA@meta.data$assign.ident)
        row.names(reg_table) = NULL
        celltype_count = table(reg_table)
        celltype_count_df = as.data.frame(celltype_count, stringsAsFactors = FALSE)
        celltype_dominate = sapply(unique(celltype_count_df$Cluster),function(x){
          celltype_count_df_cluster = celltype_count_df[celltype_count_df$Cluster == x,]
          celltype_dominate = celltype_count_df_cluster[which.max(celltype_count_df_cluster$Freq), "CelltypeAnnotation"]
        })
        reg_table_unique = data.frame(Cluster = names(celltype_dominate), CelltypeAnnotation = celltype_dominate)
      }else{
        if(length(cluster) == 1 && cluster == "unknown"){
          reg_table_unique = data.frame(Cluster = "unknown", CelltypeAnnotation = "unknown")
        }else{
          celltype = RNA$assign.ident[names(Idents(RNA))[Idents(RNA) %in% cluster]]
          celltype_count = sort(table(celltype), decreasing = TRUE)
          celltype_dominate = names(celltype_count)[1]
          reg_table_unique = data.frame(Cluster = paste(cluster, collapse = ","), CelltypeAnnotation = celltype_dominate)
        }
      }
    }else{
      reg_table_unique = data.frame(Cluster = unique(cluster_tf_df$Cluster), CelltypeAnnotation = "unknown")
    }
    reg_df <- merge(reg_table_unique, cluster_tf_df) 
    write.table(reg_df,paste0(project, ".PredictedTFTop", top.tf, ".txt"), col.names = TRUE, row.names = FALSE, sep = "\t", quote = FALSE)
  
    tfList <- as.list(as.data.frame(cluster_tf_list_filter_tf, stringsAsFactors = FALSE))
    names(tfList) = names(cluster_tf_list_filter_tf)
  }
  return(tfList)
}


RunLISALocal <- function(genes, project, organism = "GRCh38", conda.dir = "", lisa.envname = "lisa")
{
  if(organism == "GRCh38"){
    species = "hg38"
  }
  if(organism == "GRCm38"){
    species = "mm10"
  }
  
  cluster_markers_list <- split(genes, genes$cluster)
  
  outputDir <- paste0(project, ".lisa")
  if (!file.exists(outputDir)) dir.create(path=outputDir)  
  
  for(i in names(cluster_markers_list))
  {
    cluster_marker = as.matrix(cluster_markers_list[[i]]$gene)
    lisaDir <- paste0(outputDir, "/",i)
    if (!file.exists(lisaDir)) dir.create(path=lisaDir)
    write.table(cluster_marker,paste0(lisaDir,'/', i, ".txt"),sep='\t',col.names = F, row.names = F,quote = FALSE)
  }
  
  message("Start to run Lisa.")
  for(i in names(cluster_markers_list))
  {
    cmd = paste0(". ", conda.dir, "/etc/profile.d/conda.sh && conda activate ", lisa.envname, " && lisa model --method='all' --web=False --new_rp_h5=None --new_count_h5=None --species ",species, " --epigenome '['DNase','H3K27ac']' --cluster=False --covariates=False --random=True --prefix ",i,".txt"," --background=None --stat_background_number=500 --threads 8 ",outputDir,'/', i,'/',i,".txt")
    system2("/bin/bash", args = c("-c", shQuote(cmd)))
    message(paste0("Lisa in cluster ", i, " is done!"))
  }
  system("rm *.yml *.profile *.model")
  message("Lisa is done.")
  
  tf_all = NULL
  for(i in names(cluster_markers_list))
  {
    fileName = paste0(outputDir,'/', i,'/',i, '.txt_chipseq_cauchy_combine_dedup.csv')
    if (file.exists(fileName))
    {
      tf_p = read.csv(fileName)
      rownames(tf_p) = tf_p$TF
      tf = as.vector(sort(tf_p$TF))
      tf_temp = data.frame(tf_p[tf,2])
      colnames(tf_temp) = paste0(i)
      tf_all = c(tf_all,list(tf_temp))
    }
  }
  
  tf_all <- do.call(cbind, tf_all)
  tf_all_log10 = -log10(tf_all)
  if(organism == "GRCm38"){
    tf = Hmisc::capitalize(tolower(tf))
  }
  rownames(tf_all_log10)=tf
  write.table(tf_all_log10,paste0(project,'_lisa.txt'),sep='\t',quote = F)
  return(tf_all_log10)
}

RunLISAWeb <- function(genes, project, organism = "GRCh38")
{
  if(organism == "GRCh38"){
    species = "hg38"
  }
  if(organism == "GRCm38"){
    species = "mm10"
  }
  
  cluster_markers_list <- split(genes, genes$cluster)
  
  outputDir <- paste0(project, ".lisa")
  if (!file.exists(outputDir)) dir.create(path=outputDir)  
  
  res_zip_url_list = rep("zip", length(cluster_markers_list))
  names(res_zip_url_list) = names(cluster_markers_list)
  for(i in names(cluster_markers_list))
  {
    cluster_marker = cluster_markers_list[[i]]$gene
    if(length(cluster_marker) <= 500){
      cluster_marker_str = paste(cluster_marker, collapse = "\n")
    }else{
      cluster_marker_str = paste(cluster_marker[1:500], collapse = "\n")
    }
    Sys.sleep(3)
    res_zip_url = PostForm(geneset = cluster_marker_str, jobname = paste0("cluster_", i))
    message(paste0(Sys.time(),": Lisa in cluster ", i, " was posted successfully."))
    res_zip_url_list[i] = res_zip_url
  }
  
  Sys.sleep(300)
  
  for(i in names(res_zip_url_list)){
    lisaRes <- paste0(outputDir, "/",i, ".zip")
    lisaDir <- paste0(outputDir, "/",i)
    DownloadResult(res_zip_url_list[i], lisaRes)
    if(file.exists(lisaDir)){
      system(paste0("rm -r ", lisaDir))
    }
    unzip(lisaRes, exdir = lisaDir)
    file.remove(lisaRes)
    message(paste0(Sys.time(),": Lisa in cluster ", i, " is done."))
  }
  
  tf_all=NULL
  for(i in names(cluster_markers_list))
  {
    lisaDir <- paste0(outputDir, "/",i)
    fileName = list.files(lisaDir, "*chipseq_cauchy_combine_dedup.csv")
    fileName = file.path(lisaDir, fileName)
    if (file.exists(fileName))
    {
      tf_p=read.csv(fileName)
      rownames(tf_p)=tf_p$TF
      tf=as.vector(sort(tf_p$TF))
      tf_temp=data.frame(tf_p[tf,2])
      colnames(tf_temp)=paste0(i)
      tf_all=c(tf_all,list(tf_temp))
    }
  }
  
  tf_all <- do.call(cbind, tf_all)
  tf_all_log10=-log10(tf_all)
  if(organism == "GRCm38"){
    tf = Hmisc::capitalize(tolower(tf))
  }
  rownames(tf_all_log10)=tf
  write.table(tf_all_log10,paste0(project,'_lisa.txt'),sep='\t',quote = F)
  return(tf_all_log10)
}

PostForm <- function(geneset, jobname)
{
  url = "http://lisa.cistrome.org"
  
  myheaders = c('User-Agent'='Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_3) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.5 Safari/605.1.15',
                'Accept'='text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
                'Accept-Encoding'='gzip, deflat',
                'Accept-Language'='en-us')
  resp1 = httr::GET(url = url, httr::add_headers(.headers = myheaders))
  resp1_text = httr::content(resp1, "text")
  
  csrf_pattern = "<input id=\"csrf_token\" name=\"csrf_token\" type=\"hidden\" value=\"(.*)\">"
  csrf = stringr::str_match(resp1_text, csrf_pattern)[1,2]
  
  form_post = list(
    csrf_token = csrf,
    species = "hg38",
    genes = geneset,
    name = jobname
  )
  
  resp2 = httr::POST(url, httr::add_headers(.headers = myheaders), body = form_post, encode="form")
  
  resp3 = httr::GET(resp2$url)
  resp3_text = httr::content(resp3, "text")
  
  zip_pattern = "<a class=\"nav-link\" href=\"(.*).zip\">"
  res_zip = stringr::str_match(resp3_text, zip_pattern)[1,2]
  res_zip_url = paste0(url, res_zip, ".zip")
  return(res_zip_url)
}

DownloadResult <- function(resurl, destfile)
{
  resp4 = httr::GET(resurl)
  while(resp4$status_code != 200){
    Sys.sleep(60)
    resp4 = httr::GET(resurl)
  }
  
  download.file(resurl, destfile)
}


RunRABIT <- function(genes, project, rabit.path, organism = "hsa")
{
  if(organism == "hsa"){
    genes$entrezid = MAGeCKFlute::TransGeneID(genes$gene, fromType = "Symbol", toType = "Entrez",
                                 organism = organism, ensemblHost = "www.ensembl.org")
  }else{
    marker_mouse = genes$gene
    human = biomaRt::useMart("ensembl", dataset = "hsapiens_gene_ensembl")
    mouse = biomaRt::useMart("ensembl", dataset = "mmusculus_gene_ensembl")
    marker_human = biomaRt::getLDS(attributes = c("mgi_symbol"), filters = "mgi_symbol", values = marker_mouse ,mart = mouse, attributesL = "hgnc_symbol", martL = human, uniqueRows=T)
    genes = merge(genes, marker_human, by.x = "gene", by.y = "MGI.symbol")
    genes$entrezid = MAGeCKFlute::TransGeneID(genes$HGNC.symbol, fromType = "Symbol", toType = "Entrez",
                                 organism = "hsa", useBiomart = FALSE,
                                 ensemblHost = "www.ensembl.org")
  }

  genes = na.omit(genes)
  cluster_markers_list <- split(genes, genes$cluster)
  
  outputDir <- paste0(project, ".RABIT")
  if (!file.exists(outputDir)){
    dir.create(path=outputDir)
  }else{
    system(paste0("rm -rf ", outputDir))
    dir.create(path=outputDir)
  }
  for(i in names(cluster_markers_list)){    
    cluster_marker_logfc <- cluster_markers_list[[i]][,c("p_val","entrezid")]
    cluster_marker_logfc <- cluster_marker_logfc[which(!duplicated(cluster_marker_logfc$entrezid)),]
    row.names(cluster_marker_logfc) <- cluster_marker_logfc[,'entrezid']
    cluster_marker_logfc <- data.frame(logfc = cluster_marker_logfc[,1], row.names = cluster_marker_logfc[,2])
    write.table(cluster_marker_logfc, paste0(project, ".RABIT/", i, ".txt"), sep = "\t", col.names = TRUE, row.names = TRUE, quote = FALSE)
  }
  
  message("Start to run Rabit.")
  for(i in names(cluster_markers_list)){
    system(paste0("Rabit -x ", rabit.path, "/interaction.Cistrome -y ", project, ".RABIT/", i,".txt -b ", rabit.path, "/interaction.Cistrome.background -o ", project, ".RABIT/", i, ".output"))
    message(paste0("Rabit in cluster ", i, " is done!"))
  }
  message("Rabit is done.")
  
  rabit_cluster_list <- list.files(paste0(project, ".RABIT"), pattern = "output.FDR")
  rabit_cluster_list <- unlist(strsplit(rabit_cluster_list, split = ".output.FDR"))

  notf_clusters = c()
  for(m in 1:length(rabit_cluster_list)){
    i <- rabit_cluster_list[m]
    out_fdr <- read.table(paste0(project, ".RABIT/",i,".output.FDR"),header = TRUE, row.names = 1, sep = "\t", check.names = FALSE)
    if(nrow(out_fdr) == 0){
      notf_clusters = c(notf_clusters, m)
      next()
    } else{
      out_fdr <- as.data.frame(t(out_fdr))
      out_fdr$gene <- unlist(strsplit(row.names(out_fdr), split = "@"))[seq(1,2*nrow(out_fdr),2)]
      out_fdr_max <- aggregate(logfc ~ gene, data = out_fdr, min)
      colnames(out_fdr_max)[2] <- i
      break()
    }
  }
  
  if(exists("out_fdr_max")){
    if(length(rabit_cluster_list) > 1){
      if(is.null(notf_clusters)){
        start_ind = 2
      }else if(notf_clusters[length(notf_clusters)] + 1 <= length(rabit_cluster_list)){
        start_ind = notf_clusters[length(notf_clusters)] + 1
      }
      for(m in start_ind:length(rabit_cluster_list)){
        i = rabit_cluster_list[m]
        out_fdr_cur <- read.table(paste0(project, ".RABIT/", i, ".output.FDR"), header = TRUE, row.names = 1, sep = "\t", check.names = FALSE)
        if(nrow(out_fdr_cur) == 0){
          notf_clusters = c(notf_clusters, m)
          next()
        }else{
          out_fdr_cur <- as.data.frame(t(out_fdr_cur))
          out_fdr_cur$gene <- unlist(strsplit(row.names(out_fdr_cur), split = "@"))[seq(1,2*nrow(out_fdr_cur),2)]
          out_fdr_max_cur <- aggregate(logfc ~ gene, data = out_fdr_cur, min)
          colnames(out_fdr_max_cur)[2] <- i
          out_fdr_max <- merge(out_fdr_max,out_fdr_max_cur)
        }
      }
      if(!is.null(notf_clusters)){
        message(paste0("There are no significant TFs identified in Cluster ",paste(rabit_cluster_list[notf_clusters], collapse = ", "),"."))
      }
    }
    
    if(organism == "hsa"){
      out_fdr_max$gene <- unlist(strsplit(out_fdr_max$gene, split = "\\."))[seq(2,2*nrow(out_fdr_max),2)]
    }else{
      out_fdr_max$human.gene <- unlist(strsplit(out_fdr_max$gene, split = "\\."))[seq(2,2*nrow(out_fdr_max),2)]
      tf_mouse = biomaRt::getLDS(attributes = c("hgnc_symbol"), filters = "hgnc_symbol", values = out_fdr_max$human.gene ,mart = human, attributesL = "mgi_symbol", martL = mouse, uniqueRows=T)
      out_fdr_max = merge(out_fdr_max, tf_mouse,by.x = "human.gene", by.y = "HGNC.symbol")
      out_fdr_max$gene = out_fdr_max[,ncol(out_fdr_max)]
      out_fdr_max = out_fdr_max[,-c(1,ncol(out_fdr_max))]
      out_fdr_max = out_fdr_max[!duplicated(out_fdr_max$gene),]
    }
    rownames(out_fdr_max) <- out_fdr_max$gene
    if(ncol(out_fdr_max) > 2){
      out_fdr_max <- out_fdr_max[,-1]
      out_fdr_max_log <- -log10(out_fdr_max)
    }else if(ncol(out_fdr_max) == 2){
      tf_cluster = colnames(out_fdr_max)[2]
      out_fdr_max <- data.frame(cluster = out_fdr_max[,2], row.names = out_fdr_max[,1])
      colnames(out_fdr_max) = tf_cluster
      out_fdr_max_log <- -log10(out_fdr_max)
    }
    write.table(out_fdr_max_log,paste0(project,'_rabit.txt'),sep='\t',quote = F)
    return(out_fdr_max_log)
  }else{
    stop("There are no significant TFs for all the clusters.")
    tfList <- NULL
  }
}
