#' Estimate gene activity from peak using regulatory potential model
#'
#' Calculate gene score according to scATAC peak count.
#'
#' @docType methods
#' @name ATACCalculateGenescore
#' @rdname ATACCalculateGenescore
#' 
#' @param inputMat Input unnormalized count matrix, with peaks as rows and cells as columns. Rownames should be like "chromosome_peakstart_peakend",
#' for example "chr10_100020591_100020841". It can be a sparse matrix or a dense matrix. But we recommend to use the sparse one, because 
#' this will result in significant memory and speed savings.
#' @param project Output project name for the gene sore result.Default is "MAESTRO.scATAC".
#' @param organism Organism for the dataset. Only support "GRCh38" and "GRCm38". Default is "GRCh38".
#' @param decaydistance Gene score decay distance, could be optional from 1kb (promoter-based regulation) 
#' to 10kb (enhancer-based regulation). Default is 10000.
#' @param model The RP model to use to calaculate gene score, Simple or Enhanced 
#' For each gene, simple model summarizes the impact of all regulatory elements within the up/dowm-stream of TSS. 
#' On the basis of simple model, enhanced model includes the regulatory elements within the exon region, 
#' and also excludes the regulatory elements overlapped with another gene (the promoter and exon of a nearby gene). 
#' See the MAESTRO paper for more details. Default is Enhanced.
#'
#' @author Dongqing Sun, Changxin Wan, Chenfei Wang
#'
#' @return Sparse matrix of regulatory potential generated by MAESTRO. With genes as rows and cells as columns,
#' and gene RP score as values.
#'
#'
#' @examples
#' data(pbmc.ATAC)
#' pbmc.ATAC.RP.res <- ATACCalculateGenescore(inputMat = pbmc.ATAC)
#'
#' @importFrom reticulate source_python
#' @importClassesFrom Matrix dgCMatrix
#' @export 
#' 

ATACCalculateGenescore <- function(inputMat, project = "MAESTRO.scATAC", organism = "GRCh38", 
                                   decaydistance = 10000, model = "Enhanced"){
  source_python(paste(system.file(package="MAESTRO"), "ATACCalculateGenescore.py", sep="/"))

  peaks_list = rownames(inputMat)
  peaks_list = gsub(pattern = "\\W", replacement = "_", x = peaks_list)
  if(class(inputMat) != "dgCMatrix"){
    inputMat = as(as.matrix(inputMat), "dgCMatrix")
  }
  
  if(model == "Simple"){
    if(organism == "GRCh38"){
      data(GRCh38.refgenes.genescore.simple)
      refgenes.genescore = GRCh38.refgenes.genescore.simple
    }else{
      data(GRCm38.refgenes.genescore.simple)
      refgenes.genescore = GRCm38.refgenes.genescore.simple
    }
    genes_list = refgenes.genescore[,4]
    refgenes.genescore = refgenes.genescore[,-4]
    
    rp_result = calculate_RP_score(cell_peaks = inputMat, peaks_list = peaks_list, gene_bed_df = refgenes.genescore, 
                                   genes_list = genes_list, decay = decaydistance, model = model)
    
  }
  
  if(model == "Enhanced"){
    if(organism == "GRCh38"){
      data(GRCh38.refgenes.genescore.adjusted)
      refgenes.genescore = GRCh38.refgenes.genescore.adjusted
    }else{
      data(GRCm38.refgenes.genescore.adjusted)
      refgenes.genescore = GRCm38.refgenes.genescore.adjusted
    }
    rp_result = calculate_RP_score(cell_peaks = inputMat, peaks_list = peaks_list, gene_bed_df = refgenes.genescore, 
                                   genes_list = NULL, decay = decaydistance, model = model)
  }
  rp_matrix = rp_result[[1]]
  rownames(rp_matrix) = rp_result[[2]]
  colnames(rp_matrix) = colnames(inputMat)
  
  return(rp_matrix)
}

